util/localpool/localpool.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package localpool
+
+import (
+	"runtime"
+	_ "unsafe" // required by go:linkname
+)
+
+// LocalPool is an thread local object pool.
+// It's similar to sync.Pool but has some difference
+// - It contains multiple local objects instead of one.
+// - It doesn't have any lock.
+// - It never get GCed.
+type LocalPool struct {
+	sizePerProc int
+	slots       []*slot
+	newFn       func() interface{}
+	resetFn     func(obj interface{})
+}
+
+type slot struct {
+	objs    []interface{}
```

Comment:
maybe we can try https://github.com/mmatczuk/go_generics for this to reduce cast interface cost? it seems very suitable if the pool alway hold one type's data... and it seems many project use this lib - -? https://github.com/cockroachdb/cockroach/tree/master/pkg/util/interval/generic